File: rinahira/php/src/Money.php
<?php

declare(strict_types=1);

namespace app;

class Money implements Expression
{
    /**
     * コンストラクタ
     *
     * @param integer $amount
     * @param string $currency
     */
    public function __construct(public int $amount, protected string $currency)
    {
    }

    public function times(int $multiplier): Expression
    {
        return new Money($this->amount * $multiplier, $this->currency);
    }

    public function plus(Expression $addend): Expression
    {
        return new Sum($this, $addend);
    }

    /**
     * 換算
     *
     * @param string $to
     * @return Money
     */
    public function reduce(Bank $bank, string $to): Money
    {
        $rate = $bank->rate($this->currency(), $to);
        return new Money($this->amount / $rate, $to);
    }

    /**
     * 通貨取得
     * @return string
     */
    public function currency(): string
    {
        return $this->currency;
    }

    /**
     * 等価性比較
     *
     * @param Object $object
     * @return boolean
     */
    public function equals(Object $object): bool
    {
        $money = $object;
        return $this->amount === $money->amount && $this->currency() === $money->currency;
    }

    public function toString(): string
    {
        return $this->amount . " " . $this->currency();
    }
    /**
     * Dollarクラスファクトリメソッド
     * @param integer $amount
     * @return Money
     */
    static function dollar(int $amount): Money
    {
        return new Money($amount, "USD");
    }

    /**
     * Francクラスファクトリメソッド
     *
     * @param integer $amount
     * @return Money
     */
    static function franc(int $amount): Money
    {
        return new Money($amount, "CHF");
    }
}

File: shino/src/main/java/org/codepolaris/tdd/chapter1/Money.java
package org.codepolaris.tdd.chapter1;

import org.codepolaris.tdd.chapter1.utils.Currency;

import java.util.Objects;

public class Money implements Expression {
  protected int amount;
  protected String currency;

  Money(int amount, String currency) {
    this.amount = amount;
    this.currency = currency;
  }

  public Expression times(int multiplier) {
    return new Money(amount * multiplier, currency());
  }

  public Expression plus(Expression addend) {
    return new Sum(this, addend);
  }

  @Override
  public Money reduce(Bank bank, String to) {
    int rate =bank.rate(currency, to);
    return new Money(amount / rate, to);
  }

  public String currency() {
    return currency;
  }

  public static Money dollar(int amount) {
    return new Money(amount, Currency.USD);
  }

  public static Money franc(int amount) {
    return new Money(amount, Currency.CHF);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o instanceof Money money) {
      return amount == money.amount && currency().equals(money.currency());
    }
    return false;
  }

  @Override
  public int hashCode() {
    return Objects.hash(amount);
  }

  @Override
  public String toString() {
    return amount + " " + currency;
  }
}
